File: backend/src/main/java/com/smartmoneymanager/backend/service/BillService.java
package com.smartmoneymanager.backend.service;

import java.util.List;

import com.smartmoneymanager.backend.dto.request.CreateBillRequest;
import com.smartmoneymanager.backend.dto.request.UpdateBillRequest;
import com.smartmoneymanager.backend.dto.response.BillResponse;
import com.smartmoneymanager.backend.entity.enums.BillPaymentStatus;

public interface BillService {

    List<BillResponse> getBills(Long userId, BillPaymentStatus status);

    /** Unpaid bills (PENDING or OVERDUE), soonest due date first — used for the dashboard's upcoming-bills widget. */
    List<BillResponse> getUpcomingBills(Long userId, int limit);

    BillResponse getBill(Long userId, Long billId);

    BillResponse createBill(Long userId, CreateBillRequest request);

    BillResponse updateBill(Long userId, Long billId, UpdateBillRequest request);

    void deleteBill(Long userId, Long billId);

    BillResponse markAsPaid(Long userId, Long billId);
}

File: frontend/src/services/billService.js
import apiClient from './apiClient.js';

export async function getBills(status) {
  const { data } = await apiClient.get('/bills', { params: status ? { status } : undefined });
  return data.data;
}

export async function getUpcomingBills(limit = 5) {
  const { data } = await apiClient.get('/bills/upcoming', { params: { limit } });
  return data.data;
}

export async function createBill(payload) {
  const { data } = await apiClient.post('/bills', payload);
  return data.data;
}

export async function updateBill(id, payload) {
  const { data } = await apiClient.put(`/bills/${id}`, payload);
  return data.data;
}

export async function deleteBill(id) {
  await apiClient.delete(`/bills/${id}`);
}

export async function markAsPaid(id) {
  const { data } = await apiClient.post(`/bills/${id}/pay`);
  return data.data;
}
